Hw5

VIXER13 opened the pull request.

VIXER13 committed:
Реализован аллокатор

VIXER13 committed:
template friends

VIXER13 committed:
Готовый класс ShMap, но пока без блокировок

VIXER13 committed:
Добавил семафоры и исключения.

VIXER13 committed:
Тест класса ShMap

VIXER13 committed:
Класс ShMap

VIXER13 committed:
CMakeLists.txt
Merge branch 'master' of https://github.com/VIXER13/AdvCpp-2020.1 into hw5

VIXER13 committed:
CMakeLists.txt

smirnov-vs commented on examples/hw5/src/main.cpp:
```diff
@@ -0,0 +1,94 @@
+}
+
+int main() {
+    shmem::ShMap<size_t, size_t> sequence_map(1024*1024*1024), parallel_map(1024*1024*1024);
```
неименованные константы

VIXER13 replied:
Именовал

smirnov-vs commented on libraries/shmem/src/shared_memory.cpp:
```diff
@@ -0,0 +1,26 @@
+    if (mmap == MAP_FAILED) {
+        throw ShmemException{"mmap failed"};
+    }
+    return {mmap, ShmemUniquePtrDeleter{size}};
```
думаю, можно заменить deleter лямбдой, тип данных тогда вроде будет просто std::function

VIXER13 replied:
Я считаю, что это не очень хорошее решение по целому ряду причин:
Во-первых при использовании std::function указатель будет очень много весить (40 байт, против 16 в моём случае).
Во-вторых, выделение класса делитера позволяет специализировать такие указатели исключительно под шареную память и их нельзя будет спутать с другими указателями у которых делитер std::function.
В-третьих, это позволяет навязать дополнительную логику не делая дополнительных обёрток (можно узнать объём буффера, на которую указывает указатель).
Ну и в-четвёртых, хоть для делитера это не столь актуальный фактор, всё же вызов std::function заметно проигрывает прямому вызову из-за косвенной адресации.

smirnov-vs commented on libraries/shmem/src/semaphore.cpp:
```diff
@@ -0,0 +1,40 @@
+    }
+}
+
+SemaphoreLock::SemaphoreLock(Semaphore& semaphore) :
```
можно использовать lock_guard если методы переименовать

VIXER13 replied:
Не догадался так сделать. Удалил SemaphoreLock.

smirnov-vs commented on libraries/shmem/include/linear_allocator.hpp:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    void deallocate(pointer p, size_type n) {
+        ::memset(p, 0, n * sizeof(T)); // Считаем, что работаем с POD-типами, поэтому просто обнуляем их
```
это скорее действие метода destroy

VIXER13 replied:
destroy помечен как deprecated в C++17 и удалён из 20-го стандарта
https://en.cppreference.com/w/cpp/memory/allocator/destroy

smirnov-vs replied:
`allocator_traits::destroy`

VIXER13 replied:
Переделал функцию deallocate. Теперь она чистит буффер в том случае, когда мы удаляем из него абсолютно всё, а destroy оставил на совесть allocator_traits

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,358 @@
+
+template<class Key, class T, class Compare>
+bool operator==(const ShMap<Key, T, Compare>& left, const ShMap<Key, T, Compare>& right) {
+    SemaphoreLock lock_left (const_cast<ShMap<Key, T, Compare>&>(left ).getSemaphore()),
```
почему бы не сделать метод const?

VIXER13 replied:
Это дружественная функция, а не метод

smirnov-vs replied:
`.getSemaphore()` этот

VIXER13 replied:
Я уже заменил метод указателем на семафор и сделал его mutable, чтобы не мучиться с const_cast'ом.

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,358 @@
+    template <class... Args>
+    iterator emplace_hint(const_iterator hint, Args&&... args) {
+        SemaphoreLock lock(getSemaphore());
+        return getMap().emplace_hint(hint, args...);
```
rvalue станут lvalue

VIXER13 replied:
Поправил

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,358 @@
+    template<class P>
+    std::pair<iterator,bool> insert(P&& value) {
+        SemaphoreLock lock(getSemaphore());
+        return getMap().insert(value);
```
скопирует

VIXER13 replied:
Поправил

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,358 @@
+        memory_(shmem::makeShmemUniquePtr(reserve_bytes)) {
+        allocator_type alloc(memory_, sizeof(map_type) + sizeof(Semaphore));
+        map_type* map = &getMap();
+        map = new (map) map_type{alloc};
```
обычно это метод construct аллокатора

VIXER13 replied:
Добавил метод, только теперь из-за него конструктор ShMap какой-то костыльный получился, не знаю как правильней сделать

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,358 @@
+
+    ShMap(const size_t reserve_bytes = sizeof(map_type) + sizeof(Semaphore)) :
+        memory_(shmem::makeShmemUniquePtr(reserve_bytes)) {
+        allocator_type alloc(memory_, sizeof(map_type) + sizeof(Semaphore));
```
что если reserve_bytes передадут?

VIXER13 replied:
Наверное это не очень хорошее решение с моей стороны, задать минимально возможный объём памяти, который требуется для создания контейнера в параметр по-умолчанию.

А вообще, если объём будет меньше минимального, то при создании аллокатора будет выкинуто исключение, а если больше, то всё будет ок.

P.S. скорее всего эта логика немного поменяется после того как я разберусь с методом construct.

VIXER13 committed:
Удалил SemaphoreLock. Поправил методы с rvalue. Именовал константы

VIXER13 committed:
У аллокатора добавил метод construct.
Поменял конструктор у ShMap. Добавил два поля с указателями на семафор и std::map.

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,341 @@
+
+    T& at(const Key& key) {
+        std::lock_guard<Semaphore> lock(*semaphore_);
+        return map_->at(key);
```
кстати лок отпустит, а вот пользователь получит ссылку на шареную память без лока уже, будет гонка

VIXER13 replied:
Это уже по идее на совести пользователя, в моей компетенции дать ему корректную ссылку при обращении, а что с ней станет через секунду я знать не могу

smirnov-vs replied:
отдавай по значению

VIXER13 replied:
По значению, конечно, можно, но ведь std::map так же не даёт никаких гарантий в данной ситуации и уже пользователю приходится как-то выкручиваться, чтобы избежать гонки

VIXER13 replied:
Мне кажется здесь, в качестве безопасной альтернативы, можно просто добавить дополнительный метод и тогда все будут при своём

VIXER13 replied:
Мне всё-таки не очень нравится идея возвращать по значению, поэтому в качестве компромисса добавил atSafe и перегрузил operator() в качестве альтернативы operator[], которые возвращают по значению

smirnov-vs commented on libraries/shmem/include/linear_allocator.hpp:
```diff
@@ -0,0 +1,85 @@
+    using reference       = T&;
+    using const_reference = const T&;
+    template<typename U>
+	struct rebind {
```
нигде не используется?

VIXER13 replied:
Я просто полностью повторил все typedef'ы std::allocator. Но сейчас в чатике активно обсуждали std::allocator_traits, возможно я зря это всё сделал.

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,341 @@
+        shmem::LinearAllocator<Semaphore> semaphore_alloc{memory_};
+        semaphore_ = semaphore_alloc.allocate(1);
+        semaphore_alloc.construct(semaphore_);
+        shmem::LinearAllocator<map_type> map_alloc{semaphore_alloc};
```
здесь думаю достаточно только аллокатора для мапы, ведь именно для нее мы его делаем, так как там сложная логика внутри
хедер памяти можно инициализировать прямо тут вручную

VIXER13 replied:
Так и не понял как это можно адекватно переделать, поэтому оставил как есть, только метод construct использую из allocator_traits

smirnov-vs reviewed (commented):
обсудим после лекции

smirnov-vs commented on libraries/shmem/containers/shmap.hpp:
```diff
@@ -0,0 +1,341 @@
+
+template<class Key, class T, class Compare>
+void swap(ShMap<Key, T, Compare>& left, ShMap<Key, T, Compare>& right) {
+    std::lock_guard<Semaphore> lock_left(*left.semaphore_), lock_right(*right.semaphore_);
```
кстати можно заменить на scoped_lock

VIXER13 replied:
Точно, иначе бы на самопроверке случился бы deadlock

VIXER13 committed:
scoped lock

VIXER13 committed:
Удалил лишнее

VIXER13 committed:
string test

VIXER13 committed:
Merge https://github.com/VIXER13/AdvCpp-2020.1 into hw5
